source/logic.rst

```diff
@@ -0,0 +1,492 @@
+that work at a high level of generality.
+
+For another example of a hidden universal quantifier,
+mathlib defines a predicate ``monotone``,
+which says that a function is nondecreasing in its arguments:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    example (f : ℝ → ℝ) (h : monotone f) :
+      ∀ {a b}, a ≤ b → f a ≤ f b := h
+    -- END
+
+Proving statements about monotonicity
+involves using ``intros`` to introduce two variables,
+say, ``a`` and ``b``, and the hypothesis ``a ≤ b``.
+To *use* a monotonicity hypothesis,
+you can apply it to suitable arguments and hypotheses,
+and then apply the resulting expression to the goal.
+Or you can apply it to the goal and let Lean help you
+work backwards by displaying the remaining hypotheses
+as new subgoals.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    example (mf : monotone f) (mg : monotone g) :
+      monotone (λ x, f x + g x) :=
+    begin
+      intros a b aleb,
+      apply add_le_add,
+      apply mf aleb,
+      apply mg aleb
+    end
+    -- END
+
+When a proof is this short, it is often convenient
+to give a proof term instead.
+The ``intros`` command corresponds to a lambda,
+and the remaining term consists of applications.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    example (mf : monotone f) (mg : monotone g) :
+      monotone (λ x, f x + g x) :=
+    λ a b aleb, add_le_add (mf aleb) (mg aleb)
+    -- END
+
+Here is a useful trick: if you start writing
+the proof term ``λ a b aleb, _`` using
+an underscore where the rest of the
+expression should go,
+Lean will flag an error,
+indicating that it can't guess the value of that expression.
+If you check the Lean Goal window in VS Code or
+hover over the squiggly error marker,
+Lean will show you the goal that the remaining
+expression has to solve.
+
+Try proving these, with either tactics or proof terms:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    example {c : ℝ} (mf : monotone f) (nnc : 0 ≤ c) :
+      monotone (λ x, c * f x) :=
+    sorry
+
+    example (mf : monotone f) (mg : monotone g) :
+      monotone (λ x, f (g x)) :=
+    sorry
+    -- END
+
+Here are some more examples.
+A function :math:`f` from :math:`\Bbb R` to
+:math:`\Bbb R` is said to be *even* if
+:math:`f(-x) = f(x)` for every :math:`x`,
+and *odd* if :math:`f(-x) = -f(x)` for every :math:`x`.
+The following example defines these two notions formally
+and establishes one fact about them.
+You can complete the proofs of the others.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    def even (f : ℝ → ℝ) : Prop := ∀ x, f x = f (-x)
+    def odd (f : ℝ → ℝ) : Prop := ∀ x, f x = - f (-x)
+
+    example (ef : even f) (eg : even g) : even (λ x, f x + g x) :=
+    by { intro x, dsimp, rw [ef, eg] }
```

Comment:
Should we mention that this `dsimp` (or some `change`) is *needed* here because of `rw`. This is also a place where a `calc` block would read much nicer, including an explicit unfolding line:
```lean
λ x,
calc (λ x, f x + g x) x = f x + g x : rfl
... = f (-x) + g (-x) : by rw [ef, eg]
```
or the same in tactic mode of course.